File: github_code_index_sort/cpp/3/40.c
//**************************************************************
//                  Topological Sort
//**************************************************************

#ifndef TOPOLOGICALSORT_H
#define TOPOLOGICALSORT_H

#include "Digraph.h"
#include "DirectedCycle.h"
#include "DepthFirstOrder.h"
#include "EdgeWeightedDirectedCycle.h"
#include "EdgeWeightedDigraph.h"

class TopologicalSort
{
public:
    TopologicalSort(Digraph& g)
    {
        DirectedCycle dc(g);
        bool hasCycle = dc.HasCycle();
        if (!hasCycle)
        {
            DepthFirstOrder dfo(g);
            stack<int> reversePostOrder = dfo.ReversePostOrder();
            while (!reversePostOrder.empty())
            {
                int v = reversePostOrder.top();
                order.push_back(v);
                reversePostOrder.pop();
            }
        }
    }

    // Extend topologicalSort for Edge-weighted Digraph.
    TopologicalSort(EdgeWeightedDigraph& g)
    {
        EdgeWeightedDirectedCycle dc(g);
        bool hasCycle = dc.HasCycle();
        if (!hasCycle)
        {
            DepthFirstOrder dfo(g);
            stack<int> reversePostOrder = dfo.ReversePostOrder();
            while (!reversePostOrder.empty())
            {
                int v = reversePostOrder.top();
                order.push_back(v);
                reversePostOrder.pop();
            }
        }
    }

    ~TopologicalSort()
    {
    }

    // only graph is DAG(which has no cycle) can have topo-order.
    bool IsDAG()
    {
        return !order.empty();
    }

    // vertices in topological order.
    vector<int> Order()
    {
        return order;
    }

private:
    vector<int> order;
};

#endif

File: github_java_sort/3/40.java
package algs4;


public class Topological {
    private Iterable<Integer> order;    

    
    public Topological(Digraph G) {
        DirectedCycle finder = new DirectedCycle(G);
        if (!finder.hasCycle()) {
            DepthFirstOrder dfs = new DepthFirstOrder(G);
            order = dfs.reversePost();
        }
    }

    
    public Topological(EdgeWeightedDigraph G) {
        EdgeWeightedDirectedCycle finder = new EdgeWeightedDirectedCycle(G);
        if (!finder.hasCycle()) {
            DepthFirstOrder dfs = new DepthFirstOrder(G);
            order = dfs.reversePost();
        }
    }

    
    public Iterable<Integer> order() {
        return order;
    }

    
    public boolean hasOrder() {
        return order != null;
    }


    public static void main(String[] args) {
        String filename  = args[0];
        String delimiter = args[1];
        SymbolDigraph sg = new SymbolDigraph(filename, delimiter);
        Topological topological = new Topological(sg.G());
        for (int v : topological.order()) {
            StdOut.println(sg.name(v));
        }
    }

}
